Tests/LibWeb/test-web.cpp

```diff
@@ -4,664 +4,120 @@
+    view.set_document(Web::DOM::Document::create());
+    g_page_view = view;
 }
 
-JS_DEFINE_NATIVE_FUNCTION(TestRunnerObject::change_page)
+TESTJS_GLOBAL_FUNCTION(load_local_page, loadLocalPage)
 {
-    auto url = vm.argument(0).to_string(global_object);
+    auto name = vm.argument(0).to_string(global_object);
     if (vm.exception())
         return {};
 
-    if (g_on_page_change)
-        g_on_page_change(url);
+    // Clear the hooks
+    before_initial_load_hooks.clear();
+    after_initial_load_hooks.clear();
 
+    // Set the load URL
+    if (name.starts_with('/'))
+        next_page_to_load = URL::create_with_file_protocol(name);
+    else
+        next_page_to_load = URL::create_with_file_protocol(LexicalPath::join(g_test_root, "Pages", name).string());
     return JS::js_undefined();
 }
 
-class TestRunner {
-public:
-    TestRunner(String web_test_root, String js_test_root, Web::InProcessWebView& page_view, bool print_times)
-        : m_web_test_root(move(web_test_root))
-        , m_js_test_root(move(js_test_root))
-        , m_print_times(print_times)
-        , m_page_view(page_view)
-    {
-    }
-
-    void run();
-
-private:
-    JSFileResult run_file_test(const String& test_path);
-    void print_file_result(const JSFileResult& file_result) const;
-    void print_test_results() const;
-
-    String m_web_test_root;
-    String m_js_test_root;
-    bool m_print_times;
-
-    double m_total_elapsed_time_in_ms { 0 };
-    Test::Counts m_counts;
-
-    RefPtr<Web::InProcessWebView> m_page_view;
-
-    RefPtr<JS::Program> m_js_test_common;
-    RefPtr<JS::Program> m_web_test_common;
-};
-
-static void cleanup_and_exit()
-{
-    // Clear the taskbar progress.
-#ifdef __serenity__
-    fprintf(stderr, "\033]9;-1;\033\\");
-#endif
-    exit(1);
-}
-
-#if 0
-static void handle_sigabrt(int)
-{
-    dbgln("test-web: SIGABRT received, cleaning up.");
-    cleanup_and_exit();
-}
-#endif
-
-static double get_time_in_ms()
+TESTJS_GLOBAL_FUNCTION(after_initial_page_load, afterInitialPageLoad)
 {
-    struct timeval tv1;
-    auto return_code = gettimeofday(&tv1, nullptr);
-    VERIFY(return_code >= 0);
-    return static_cast<double>(tv1.tv_sec) * 1000.0 + static_cast<double>(tv1.tv_usec) / 1000.0;
-}
-
-template<typename Callback>
-void iterate_directory_recursively(const String& directory_path, Callback callback)
-{
-    Core::DirIterator directory_iterator(directory_path, Core::DirIterator::Flags::SkipDots);
-
-    while (directory_iterator.has_next()) {
-        auto file_path = directory_iterator.next_full_path();
-        if (Core::File::is_directory(file_path)) {
-            iterate_directory_recursively(file_path, callback);
-        } else {
-            callback(move(file_path));
-        }
-    }
-}
-
-static Vector<String> get_test_paths(const String& test_root)
-{
-    Vector<String> paths;
+    auto function = vm.argument(0);
+    if (!function.is_function())
+        return {};
```

Comment:
Not sure if it was like this before, but let's not leak an empty value here - if this should not be allowed, throw an exception instead and _then_ return an empty value. If it _should_ be ok for whatever reason, just return `js_undefined()`.

Reply:
Throwing an exception from these tends to make test-js really sad, since no one catches them.
Also, leak an empty value? isn't that just a value type? (i.e. no heap allocation)

Reply:
ah sorry, I didn't mean leak as in leak memory but as in exposing it to the non-C++ side of LibJS - i.e. when you do `var foo = afterInitialPageLoad(42)`, foo now holds an empty `JS::Value`, and doing virtually anything with that will blow up.

Reply:
Ah okay.
in that case, I'll log something and then throw